// tsparticles-plugin-sounds/cjs/Options/Classes/Sounds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sounds = void 0;
const SoundsEvent_1 = require("./SoundsEvent");
const SoundsIcons_1 = require("./SoundsIcons");
const SoundsVolume_1 = require("./SoundsVolume");
class Sounds {
    constructor() {
        this.enable = false;
        this.events = [];
        this.icons = new SoundsIcons_1.SoundsIcons();
        this.volume = new SoundsVolume_1.SoundsVolume();
    }
    load(data) {
        if (!data) {
            return;
        }
        if (data.enable !== undefined) {
            this.enable = data.enable;
        }
        if (data.events !== undefined) {
            this.events = data.events.map((t) => {
                const event = new SoundsEvent_1.SoundsEvent();
                event.load(t);
                return event;
            });
        }
        this.icons.load(data.icons);
        if (data.volume !== undefined) {
            this.volume.load(data.volume);
        }
    }
}
exports.Sounds = Sounds;
