// tsparticles-plugin-sounds/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadSoundsPlugin = void 0;
const Sounds_1 = require("./Options/Classes/Sounds");
const SoundsInstance_1 = require("./SoundsInstance");
class SoundsPlugin {
    constructor(engine) {
        this.id = "sounds";
        this._engine = engine;
    }
    getPlugin(container) {
        return new SoundsInstance_1.SoundsInstance(container, this._engine);
    }
    loadOptions(options, source) {
        if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
            return;
        }
        let soundsOptions = options.sounds;
        if (soundsOptions?.load === undefined) {
            options.sounds = soundsOptions = new Sounds_1.Sounds();
        }
        soundsOptions.load(source?.sounds);
    }
    needsPlugin(options) {
        return options?.sounds?.enable ?? false;
    }
}
async function loadSoundsPlugin(engine, refresh = true) {
    await engine.addPlugin(new SoundsPlugin(engine), refresh);
}
exports.loadSoundsPlugin = loadSoundsPlugin;
